demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+                            bufferLength, length ) );
+                status = false;
+            }
+        }
+
+        if( status == true )
+        {
+            /* Return to the beginning of the file. */
+            result = fseek( file, 0L, SEEK_SET );
+
+            if( result == -1 )
+            {
+                LogError( ( "Failed to move to beginning of file. Path: %s. Error: %s.",
+                            path, strerror( errno ) ) );
+                status = false;
+            }
+        }
+
+        if( status == true )
+        {
+            size_t written = 0;
+            /* Read the CSR into our buffer. */
+            written = fread( pBuffer, 1, length, file );
+
+            if( written != length )
+            {
+                LogError( ( "Failed reading file. Path: %s. Error: %s.", path,
+                            strerror( errno ) ) );
+                status = false;
+            }
+            else
+            {
+                *pOutWrittenLength = length;
+            }
+        }
+
+        fclose( file );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV destroyProvidedObjects( CK_SESSION_HANDLE session,
+                                     CK_BYTE_PTR * pkcsLabelsPtr,
+                                     CK_OBJECT_CLASS * pClass,
+                                     CK_ULONG count )
+{
+    CK_RV result;
+    CK_FUNCTION_LIST_PTR functionList;
+    CK_OBJECT_HANDLE objectHandle;
+    CK_BYTE * labelPtr;
+    CK_ULONG index = 0;
+
+    result = C_GetFunctionList( &functionList );
+
+    for( index = 0; index < count; index++ )
+    {
+        labelPtr = pkcsLabelsPtr[ index ];
+
+        result = xFindObjectWithLabelAndClass( session, ( char * ) labelPtr,
+                                               strlen( ( char * ) labelPtr ),
+                                               pClass[ index ], &objectHandle );
+
+        while( ( result == CKR_OK ) && ( objectHandle != CK_INVALID_HANDLE ) )
+        {
+            result = functionList->C_DestroyObject( session, objectHandle );
+
+            /* PKCS #11 allows a module to maintain multiple objects with the same
+             * label and type. The intent of this loop is to try to delete all of
+             * them. However, to avoid getting stuck, we won't try to find another
+             * object of the same label/type if the previous delete failed. */
+            if( result == CKR_OK )
+            {
+                result = xFindObjectWithLabelAndClass( session, ( char * ) labelPtr,
+                                                       strlen( ( char * ) labelPtr ),
+                                                       pClass[ index ], &objectHandle );
+            }
+            else
+            {
+                break;
+            }
+        }
+
+        if( result == CKR_FUNCTION_NOT_SUPPORTED )
```

Comment:
Can this return code come from `C_DestroyObject` function? Please add a comment about it